### PR TITLE
CORTX-33398: Added deployment query in CORTX deployment pipeline

### PR DESCRIPTION
**Pull request by shailesh-vaidya**
# Problem Statement

- Added stage in Jenkins pipeline to query deployment and print output. Used api/v2/system/topology as of now.
- Query output will be available in Jenkins console logs and also in the management-path-status.txt file as build artifacts.
- Please refer - https://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/sv_space/job/setup-cortx-cluster/215/console

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
- [x] Jenkins pipelines used for testing - https://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/sv_space/job/setup-cortx-cluster/215/

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by shailesh-vaidya** [comment]
Query output will be stored in `management-path-status.txt` and will be available as build artifacts 
```
==================== Executing api/v2/system/topology ====================

{
  "topology": {
    "cluster_id": "a09edaba-cc7a-4a37-8652-a62fc743d6b1",
    "version": "2.0.0-940",
    "nodes": [
      {
        "id": "3b404b96fc01319b5edb2e84e323595a",
        "version": "2.0.0-940",
        "services": [
          "rgw_s3"
        ],
        "type": "server_node",
        "storage_set": "storage-set-1",
        "deployment_time": "1662553046",
        "hostname": "cortx-server-1.cortx-server-headless.cortx.svc.cluster.local"
      },
      {
        "id": "6ac2b08cfe659acf910fed4d11b7df53",
        "version": "2.0.0-940",
        "services": [
          "io"
        ],
        "type": "data_node/0",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552981",
        "hostname": "cortx-data-g0-2.cortx-data-headless.cortx.svc.cluster.local",
        "storage": [
          {
            "devices": {
              "data": [
                "/dev/sdd",
                "/dev/sde"
              ],
              "metadata": [
                "/dev/sdc"
              ]
            },
            "name": "cvg-01",
            "type": "ios"
          }
        ]
      },
      {
        "id": "7ba8990a85d38508ca9bac4ff516d3ea",
        "version": "2.0.0-940",
        "services": [
          "io"
        ],
        "type": "data_node/0",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552980",
        "hostname": "cortx-data-g0-0.cortx-data-headless.cortx.svc.cluster.local",
        "storage": [
          {
            "devices": {
              "data": [
                "/dev/sdd",
                "/dev/sde"
              ],
              "metadata": [
                "/dev/sdc"
              ]
            },
            "name": "cvg-01",
            "type": "ios"
          }
        ]
      },
      {
        "id": "7d9ee991ae6ce628a0e10294bdebaa6b",
        "version": "2.0.0-940",
        "type": "ha_node",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552991",
        "hostname": "cortx-ha-headless"
      },
      {
        "id": "aa378e768a2ba5291d786b3dc48d5650",
        "version": "2.0.0-940",
        "services": [
          "io"
        ],
        "type": "data_node/1",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552978",
        "hostname": "cortx-data-g1-0.cortx-data-headless.cortx.svc.cluster.local",
        "storage": [
          {
            "devices": {
              "data": [
                "/dev/sdg",
                "/dev/sdh"
              ],
              "metadata": [
                "/dev/sdf"
              ]
            },
            "name": "cvg-02",
            "type": "ios"
          }
        ]
      },
      {
        "id": "ac2f9656e38a92990522025e1f3ccaef",
        "version": "2.0.0-940",
        "services": [
          "rgw_s3"
        ],
        "type": "server_node",
        "storage_set": "storage-set-1",
        "deployment_time": "1662553047",
        "hostname": "cortx-server-0.cortx-server-headless.cortx.svc.cluster.local"
      },
      {
        "id": "cd1e4c61d297a784245dd88106262cd3",
        "version": "2.0.0-940",
        "services": [
          "io"
        ],
        "type": "data_node/0",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552982",
        "hostname": "cortx-data-g0-1.cortx-data-headless.cortx.svc.cluster.local",
        "storage": [
          {
            "devices": {
              "data": [
                "/dev/sdd",
                "/dev/sde"
              ],
              "metadata": [
                "/dev/sdc"
              ]
            },
            "name": "cvg-01",
            "type": "ios"
          }
        ]
      },
      {
        "id": "d62dc0d2c3ee12aee9b946379519133a",
        "version": "2.0.0-940",
        "services": [
          "rgw_s3"
        ],
        "type": "server_node",
        "storage_set": "storage-set-1",
        "deployment_time": "1662553046",
        "hostname": "cortx-server-2.cortx-server-headless.cortx.svc.cluster.local"
      },
      {
        "id": "ecabccb4bced858e282f568d6bf04558",
        "version": "2.0.0-940",
        "services": [
          "agent"
        ],
        "type": "control_node",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552959",
        "hostname": "cortx-control"
      },
      {
        "id": "f488442b875fe8c4328e148f8aca7aa7",
        "version": "2.0.0-940",
        "services": [
          "io"
        ],
        "type": "data_node/1",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552982",
        "hostname": "cortx-data-g1-2.cortx-data-headless.cortx.svc.cluster.local",
        "storage": [
          {
            "devices": {
              "data": [
                "/dev/sdg",
                "/dev/sdh"
              ],
              "metadata": [
                "/dev/sdf"
              ]
            },
            "name": "cvg-02",
            "type": "ios"
          }
        ]
      },
      {
        "id": "fae544140e673b0f2a0d26cb2011516a",
        "version": "2.0.0-940",
        "services": [
          "io"
        ],
        "type": "data_node/1",
        "storage_set": "storage-set-1",
        "deployment_time": "1662552979",
        "hostname": "cortx-data-g1-1.cortx-data-headless.cortx.svc.cluster.local",
        "storage": [
          {
            "devices": {
              "data": [
                "/dev/sdg",
                "/dev/sdh"
              ],
              "metadata": [
                "/dev/sdf"
              ]
            },
            "name": "cvg-02",
            "type": "ios"
          }
        ]
      }
    ],
    "storage_sets": [
      {
        "id": "storage-set-1",
        "durability": {
          "data": "1+0+0",
          "metadata": "1+0+0"
        }
      }
    ],
    "certificates": [
      {
        "subject": {
          "countryName": "IN",
          "stateOrProvinceName": "MH",
          "localityName": "Pune",
          "organizationName": "Seagate Technology",
          "commonName": "seagate.com"
        },
        "issuer": {
          "countryName": "IN",
          "stateOrProvinceName": "MH",
          "localityName": "Pune",
          "organizationName": "Seagate Technology",
          "commonName": "seagate.com"
        },
        "not_valid_after": "2032-03-05 09:47:00",
        "not_valid_before": "2022-03-08 09:47:00",
        "serial_number": 6.419409336135354e+47,
        "version": "Version.v3",
        "name": "cortx.pem",
        "type": "SSL Certificate"
      }
    ]
  }
}

--------------- SUCCESS ---------------
```

File: solutions/kubernetes/cortx-deploy.sh [hunk 1]
```diff
@@ -183,11 +184,17 @@ function io-sanity() {
         if [ -f '$SOLUTION_CONFIG' ]; then echo "file $SOLUTION_CONFIG not available..."; exit 1; fi
     fi
 
-    add_primary_separator "\tSetting up IO Sanity Testing"
+    add_primary_separator "\tSetting up IO Path Health Check"
```
[the remark this answers is not in the document]
Comment (shailesh-vaidya) [reply solutions/kubernetes/cortx-deploy.sh:187]: People get confused with this test as QA sanity testing. Hence we are planning to use `Health Check` for all DevOps checks

File: solutions/kubernetes/management-path-check.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+#
+# Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
```
[the remark this answers is not in the document]
Comment (shailesh-vaidya) [reply solutions/kubernetes/management-path-check.sh]: Addressed, Please review again

File: solutions/kubernetes/management-path-check.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+source /var/tmp/functions.sh
+
+CORTX_MANAGER_IP=$(kubectl get svc cortx-control -o=jsonpath='{.spec.clusterIP}')
+CORTX_MANAGER_PORT=$(kubectl get svc cortx-control -o=jsonpath='{.spec.ports[?(@.name=="control-https")].port}')
+CORTX_MANAGER_ENDPOINT="https://$CORTX_MANAGER_IP:$CORTX_MANAGER_PORT"
+
+function validate_packages() {
+PACKAGE_LIST=$*
+for package in $PACKAGE_LIST; do rpm -q $package || yum install $package -y || (add_common_separator "Failed to install $package. Existing"; exit 1); done
```
[the remark this answers is not in the document]
Comment (shailesh-vaidya) [reply solutions/kubernetes/management-path-check.sh]: Addressed, Please review again

File: solutions/kubernetes/management-path-check.sh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#Function to execute API queries using curl
+function run_curl_query() {
+    add_secondary_separator "Executing $1"
+    HTTP_RESPONSE=$(curl -k -s -w "%{http_code}" $CORTX_MANAGER_ENDPOINT/$1 -o output.json)
+    jq . output.json
+    [ "$HTTP_RESPONSE" == "200" ] || (add_common_separator "ERROR: $1 execution failed. Exiting"; exit 1)
+} 
+
+#Cleanup
+rm -f $MANAGEMENT_CHECK_STATUS
```
[the remark this answers is not in the document]
Comment (shailesh-vaidya) [reply solutions/kubernetes/management-path-check.sh]: Addressed, Please review again

File: solutions/kubernetes/management-path-check.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+set -eo pipefail
+
+source /var/tmp/functions.sh
+
+CORTX_MANAGER_IP=$(kubectl get svc cortx-control -o=jsonpath='{.spec.clusterIP}')
+CORTX_MANAGER_PORT=$(kubectl get svc cortx-control -o=jsonpath='{.spec.ports[?(@.name=="control-https")].port}')
+CORTX_MANAGER_ENDPOINT="https://$CORTX_MANAGER_IP:$CORTX_MANAGER_PORT"
+
+function validate_packages() {
+PACKAGE_LIST=$*
```
[the remark this answers is not in the document]
Comment (shailesh-vaidya) [reply solutions/kubernetes/management-path-check.sh]: Addressed, Please review again

File: solutions/kubernetes/cortx-deploy.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (shailesh-vaidya) [reply solutions/kubernetes/cortx-deploy.sh:187]: There are `IO Path` (Motr, RGW, Hare) and `Management Path` (CSM, HA) hence kept work as `Path`

**Comment by shailesh-vaidya** [comment]
Tested changes with - https://eos-jenkins.colo.seagate.com/job/Release_Engineering/job/re-workspace/job/sv_space/job/setup-cortx-cluster/220/console

**Review by gauravchaudhari02 — APPROVED** [review]

**Pull request merged**